gnome-initial-setup/pages/endless-eula/gis-endless-eula-viewer.c

```diff
@@ -0,0 +1,298 @@
+
+  for (i = 0; data_dirs[i] != NULL; i++)
+    {
+      path = g_build_filename (data_dirs[i],
+                               "eos-license-service",
+                               "terms",
+                               product_name,
+                               language,
+                               "Terms-of-Use.pdf",
+                               NULL);
+
+      if (g_file_test (path, G_FILE_TEST_EXISTS))
+        return path;
+
+      g_free (path);
+    }
+
+  return NULL;
+}
+
+static char *
+find_terms_document_for_languages (const gchar *product_name,
+                                   const gchar * const *languages)
+{
+  int i;
+  gchar *path = NULL;
+
+  if (product_name == NULL)
+    return NULL;
+
+  for (i = 0; languages[i] != NULL; i++)
+    {
+      path = find_terms_document_for_language (product_name, languages[i]);
+
+      if (path != NULL)
+        return path;
+    }
+
+  /* Use "C" as a fallback. */
+  return find_terms_document_for_language (product_name, "C");
+}
+
+static GFile *
+get_terms_document (void)
+{
+  GisDriver *driver;
+  g_auto(GStrv) locale_variants = NULL;
+  const gchar *product_name;
+  const gchar *language;
+  g_autofree gchar *path = NULL;
+
+  driver = GIS_DRIVER (g_application_get_default ());
+  language = gis_driver_get_user_language (driver);
+  locale_variants = g_get_locale_variants (language);
+  product_name = gis_driver_get_product_name (driver);
+
+  path = find_terms_document_for_languages (product_name, (const gchar * const *)locale_variants);
+  if (path == NULL)
+    path = find_terms_document_for_languages ("eos", (const gchar * const *)locale_variants);
+
+  if (path == NULL)
+    {
+      g_critical ("Unable to find terms and conditions PDF on the system");
+      return NULL;
+    }
+
+  return g_file_new_for_path (path);
+}
+
+static gboolean
+gis_endless_eula_viewer_decide_policy_cb (GisEndlessEulaViewer     *self,
+                                          WebKitPolicyDecision     *decision,
+                                          WebKitPolicyDecisionType  decision_type,
+                                          gpointer                  user_data)
+{
+  g_return_val_if_fail (GIS_IS_ENDLESS_EULA_VIEWER (self), FALSE);
+
+  if (decision_type != WEBKIT_POLICY_DECISION_TYPE_NAVIGATION_ACTION)
+    return FALSE;
+
+  WebKitNavigationPolicyDecision *navigation_decision = WEBKIT_NAVIGATION_POLICY_DECISION (decision);
+  WebKitNavigationAction *action = webkit_navigation_policy_decision_get_navigation_action (navigation_decision);
+  WebKitURIRequest *request = webkit_navigation_action_get_request (action);
+  const gchar *uri = webkit_uri_request_get_uri (request);
+
+  if (g_strcmp0 (uri, self->terms_uri) == 0)
+    {
+      webkit_policy_decision_use (decision);
+      return TRUE;
+    }
+
+  webkit_policy_decision_ignore (decision);
+
+  const gchar *title = NULL;
+
+  if (g_str_has_prefix (uri, "file:"))
+    title = _("Terms of Use");
```

Comment:
I don't really think this matters. We only show Initial Setup in the locked-down kiosk so you can't get to a file to drag in. And there is no security boundary here – just trying to avoid having to deal with unintentional navigation away from the terms of use.

Reply:
Also this is the popover where no attempt is made to prevent further navigation. I think it's fine because you can just close the window if you get stuck .